app/src/main/java/org/thoughtcrime/securesms/contacts/ContactAccessor.java

```diff
@@ -0,0 +1,340 @@
+      dest.writeLong(id);
+      dest.writeString(name);
+      dest.writeTypedList(numbers);
+    }
+  }
+
+  /***
+   * If the code below looks shitty to you, that's because it was taken
+   * directly from the Android source, where shitty code is all you get.
+   */
+
+  public Cursor getCursorForRecipientFilter(CharSequence constraint,
+      ContentResolver mContentResolver)
+  {
+    final String SORT_ORDER = Contacts.TIMES_CONTACTED + " DESC," +
+                              Contacts.DISPLAY_NAME + "," +
+                              Contacts.Data.IS_SUPER_PRIMARY + " DESC," +
+                              Phone.TYPE;
+
+    final String[] PROJECTION_PHONE = {
+        Phone._ID,                  // 0
+        Phone.CONTACT_ID,           // 1
+        Phone.TYPE,                 // 2
+        Phone.NUMBER,               // 3
+        Phone.LABEL,                // 4
+        Phone.DISPLAY_NAME,         // 5
+    };
+
+    String phone = "";
+    String cons  = null;
+
+    if (constraint != null) {
+      cons = constraint.toString();
+
+      if (RecipientsAdapter.usefulAsDigits(cons)) {
+        phone = PhoneNumberUtils.convertKeypadLettersToDigits(cons);
+        if (phone.equals(cons) && !PhoneNumberUtils.isWellFormedSmsAddress(phone)) {
+          phone = "";
+        } else {
+          phone = phone.trim();
+        }
+      }
+    }
+    Uri uri = Uri.withAppendedPath(Phone.CONTENT_FILTER_URI, Uri.encode(cons));
+    String selection = String.format("%s=%s OR %s=%s OR %s=%s",
+                                     Phone.TYPE,
+                                     Phone.TYPE_MOBILE,
+                                     Phone.TYPE,
+                                     Phone.TYPE_WORK_MOBILE,
+                                     Phone.TYPE,
+                                     Phone.TYPE_MMS);
+
+    Cursor phoneCursor = mContentResolver.query(uri,
+                                                PROJECTION_PHONE,
+                                                null,
+                                                null,
+                                                SORT_ORDER);
+
+    if (phone.length() > 0) {
+      ArrayList result = new ArrayList();
+      result.add(Integer.valueOf(-1));                    // ID
+      result.add(Long.valueOf(-1));                       // CONTACT_ID
+      result.add(Integer.valueOf(Phone.TYPE_CUSTOM));     // TYPE
+      result.add(phone);                                  // NUMBER
+
+    /*
+    * The "\u00A0" keeps Phone.getDisplayLabel() from deciding
+    * to display the default label ("Home") next to the transformation
+    * of the letters into numbers.
+    */
+      result.add("\u00A0");                               // LABEL
+      result.add(cons);                                   // NAME
+
+      ArrayList<ArrayList> wrap = new ArrayList<ArrayList>();
+      wrap.add(result);
+
+      ArrayListCursor translated = new ArrayListCursor(PROJECTION_PHONE, wrap);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *translated*. Execution paths that do not contain closure statements were detected.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).